<a href="https://colab.research.google.com/github/Barbara-sudo/titanic/blob/master/Titanic_Contest_Exploration_et_nettoyage_oneHot_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Test avec un arbre de décision page 1/2
Avantage de l'arbre de décision: 
Ils sont simples à comprendre et à visualiser.
Ils nécessitent peu de préparation des données (normalisation, etc.).
Le coût d’utilisation des arbres est logarithmique.
Ils sont capables d’utiliser des données catégorielles et numériques.
Ils sont capables de traiter des problèmes multi-classe.
Modèle white box : le résultat est facile à conceptualiser et à visualiser.

A Guide to Decision Trees for Machine Learning and Data Science
Ref : https://towardsdatascience.com/a-guide-to-decision-trees-for-machine-learning-and-data-science-fe2607241956

# 1- Installer & Importer les librairies

In [0]:
import numpy as np   #librairies to support matrices and large array
import pandas as pd   #library providing high-performance, easy-to-use data structures and data analysis tools
import io
from google.colab import files
import matplotlib.pyplot as plt #
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder

**2- Importer les jeux de données**

Le jeu de données **train** correspond au jeu de données d'entraînement - Il contient les informations sur les passagers (Name, Sex, Age, SibSp, etc..) & et l'indication si ils ont survécu (1) ou non (0) dans la colonne "Survived".

Le jeu de données **test** est le jeu de données sur lequel on va prévenir les survivants et les morts

Le jeu de données **gender_submission** contient le PassengerId .
Ces jeux de données vont être utiliser pour prédire si les passengers du jeux de données train vnt surivre ou non

In [4]:
from google.colab import files
uploaded = files.upload() #permet d'uploader les fichiers présent sur ton PC
train = pd.read_csv(io.BytesIO(uploaded['train.csv']))
train.head()




Saving train.csv to train.csv


PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [0]:
#import test dataset  - 
uploaded = files.upload() #permet d'uploader les fichiers présent sur ton PC
test = pd.read_csv(io.BytesIO(uploaded['test.csv'])) #jeu de données test
test.head()




In [0]:
#import dataset  - 
uploaded = files.upload() #permet d'uploader les fichiers présent sur ton PC
gender_submission = pd.read_csv(io.BytesIO(uploaded['gender_submission.csv'])) #jeu de données test
gender_submission.head()


Saving gender_submission.csv to gender_submission.csv


PassengerId  Survived
0          892         0
1          893         1
2          894         0
3          895         0
4          896         1

**3- Explorer les données**

In [0]:
# % of survived people all genders
print("Nombre de passasgers sur le Titanic:")
Nbre_passagers = train.shape[0]
print(Nbre_passagers)

print("Nombre de personnes ayant survécues:")
all_survivors = sum(train["Survived"])
print(all_survivors)

print("% de personnes ayant survécues:")
all_gender = sum(train["Survived"])/train.shape[0] *100
print(round(all_gender,2))


Nombre de passasgers sur le Titanic:
891
Nombre de personnes ayant survécues:
342
% de personnes ayant survécues:
38.38


**Calcul du pourcentage de bonnes prédictions si l'on fait des prédictions aléatoires sachant que %survivant = 38.38%**

Soit x=1  <=> Évenement x a survécu
d'après le jeu de données train :

p(x=1) = 38.3% 

p(x=0) =61.7%

On cherche à caculer la probabilité de l'évenement : bonne prédiction pour une personne
p = p(x1=1,x2=1) + p(x1=0,x2=0) = 0.5165

P = p(x1=1,x2=1) + p(x1=0,x2=0) = 0.38^2 +0.62^2 = 0.5288
1-P = p(x1=0,x2=1) + p(x1=1,x2=0) = 0.4712

Vérif: 

P de tous les événements = 0.5288 + 0.4722

P de tous les événements = 1   ---> Ok 



In [0]:
print("Nombre d'hommes sur le Titanic:")
nb_hommes = train[train["Sex"]=="male"].shape[0]
print(nb_hommes)
print("% d'hommes sur le nombre de passagers total:")
percent_hommes = nb_hommes /train.shape[0]*100
print(round(percent_hommes,2))

print("% d'hommes sur le nombre de passagers total:")
percent_hommes = nb_hommes /train.shape[0]*100
print(round(percent_hommes,2))
print("% d'hommes parmi les survivants:")
#percent_hommes_surv = nb_hommes /train.shape[0]*100
#print(round(percent_hommes,2))
#female = sum(train[train["Sex"]=="female"]["Survived"])/train[train["Sex"]=="female"].shape[0]*100
#print(round(female,2), "% des femmes ont survécu.")
#male = sum(train[train["Sex"]=="male"]["Survived"])/train[train["Sex"]=="male"].shape[0]*100
#print(round(male,2), "% des hommes ont survécu.")


#survivors = train[train["Survived"]==1]
#percent_hommes_parmis_les_survivants = sum(survivors[survivors["Sex"]=="male"]["Survived"])/survivors.shape[0]*100
#print("Parmi les survivants :",round(percent_hommes_parmis_les_survivants,2), "% sont des hommes")


#Le sexe semble etre un paramètre important pour prédire le survivance

Variable de l'âge 
1-tester cette variable chez les hommes vs chez les femmes vs sur l'ensemble de la population

In [0]:
survivors = train[train["Survived"]== 1]
not_survivors = train[train["Survived"]== 0]
hommes_s = survivors[survivors["Sex"]== "male"]
hommes_ns = not_survivors[not_survivors["Sex"]== "male"]
femmes_s = survivors[survivors["Sex"]== "female"]
femmes_ns = not_survivors[not_survivors["Sex"]== "female"]
#plt.hist(hommes_ns["Age"])
#plt.hist(hommes_s["Age"])
plt.hist(femmes_s["Age"])
plt.hist(femmes_ns["Age"])
sum(train["Survived"])


In [0]:
train["Embarked"].unique()
train["Cabin"].unique()

#res = train["C"].str.contains("C")

#print(res)

**4-Préparer les données des jeux de données train et test pour faire un arbre de décision puis pour le tester**

(1) https://stackoverflow.com/questions/38108832/passing-categorical-data-to-sklearn-decision-tree

**According to (1) decision trees just require Integer Encoding**

>


1.   Transformer les valeur de la colonne Sex de sorte que male    --> 0   &  female --> 1

1.   Faire un integer Encoding vector pour la cabine 
tel que si cabin = NaN -->0 et si cabin != NaN -->1

2.   Transformer les lieu d'embarquement 


 







In [0]:
#4-1  Sex

# instantiate labelencoder object
le = LabelEncoder()

    #pour le jeu de données train
le = preprocessing.LabelEncoder()
le.fit(["male","female"])
train["Sex"]=le.transform(train["Sex"]) #pour le jeu de données train

train.head()

    






/usr/local/lib/python3.6/dist-packages/numpy/lib/arraysetops.py:564: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask &= (ar1 != a)


ValueError: ignored

In [0]:
#Pour le jeu de données test
le = preprocessing.LabelEncoder()
le.fit(["male","female"])
test["Sex"] = le.transform(test["Sex"])#pour le jeu de données test



>4-2 - transformer les valeur de la colonne Cabin






In [0]:
#4-2-faire un integer Encoding vector pour la cabine 
# tel que si cabin = NaN -->0 &
# et si cabin != NaN -->1

   #pour le jeu de données train
train["Cabin"].loc[~train["Cabin"].isnull()] = 1
train["Cabin"].fillna(0, inplace=True)
train["Cabin"].head()



/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:205: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)


0    0
1    1
2    0
3    1
4    0
Name: Cabin, dtype: int64

In [0]:
#Pour le jeu de donées test
test["Cabin"].loc[~test["Cabin"].isnull()] = 1
test["Cabin"].fillna(0, inplace=True)
test["Cabin"].head()

/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:205: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)


0    0
1    0
2    0
3    0
4    0
Name: Cabin, dtype: int64

In [0]:
test["Cabin"].unique()

array([0, 1])

In [0]:
#4-3  Embarked  

#pour train dataset
train["Embarked"].fillna("U", inplace=True)
train["Embarked"].unique()

le = preprocessing.LabelEncoder()
le.fit(["C","Q","S", "U"])
train["Embarked"]=le.transform(train["Embarked"])




In [0]:
#pour le test dataset
test["Embarked"].fillna("U", inplace=True)
test["Embarked"].unique()
index_EmbarkedNull = train[train["Embarked"].isnull()]
print(index_EmbarkedNull)
le = preprocessing.LabelEncoder()
le.fit(["C","Q","S", "U"])
test["Embarked"]=le.transform(test["Embarked"])


Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []


In [0]:
test["Embarked"].unique()

array([1, 2, 0])

In [0]:
#4-5 Remove NaN values from all colum
#check where are NaN
#No NaN in Pclass, Sex, SibSp, Parch, Fare, Cabin & Embarked

  # -> some NaN only one we can either replace it by the mean or remove it
train["Age"].mean()
train["Age"].fillna(train["Age"].mean(), inplace = True)
train["Age"].unique()
train["Age"].mean



<bound method Series.mean of 0      22.000000
1      38.000000
2      26.000000
3      35.000000
4      35.000000
         ...    
886    27.000000
887    19.000000
888    29.699118
889    26.000000
890    32.000000
Name: Age, Length: 891, dtype: float64>

In [0]:
#test.describe()
meanAge = test["Age"].mean() # about 100 nan --> replace by mean(age)
test[np.isnan(test["Age"])]["Age"].mean()
test["Age"].fillna(meanAge, inplace=True)

meanFare = test["Fare"].mean() # about 100 nan --> replace by mean(age)
test[np.isnan(test["Age"])]["Age"].mean()
test["Fare"].fillna(meanFare, inplace=True)




In [0]:
test["Age"].unique()

array([34.5       , 47.        , 62.        , 27.        , 22.        ,
       14.        , 30.        , 26.        , 18.        , 21.        ,
       30.27259036, 46.        , 23.        , 63.        , 24.        ,
       35.        , 45.        , 55.        ,  9.        , 48.        ,
       50.        , 22.5       , 41.        , 33.        , 18.5       ,
       25.        , 39.        , 60.        , 36.        , 20.        ,
       28.        , 10.        , 17.        , 32.        , 13.        ,
       31.        , 29.        , 28.5       , 32.5       ,  6.        ,
       67.        , 49.        ,  2.        , 76.        , 43.        ,
       16.        ,  1.        , 12.        , 42.        , 53.        ,
       26.5       , 40.        , 61.        , 60.5       ,  7.        ,
       15.        , 54.        , 64.        , 37.        , 34.        ,
       11.5       ,  8.        ,  0.33      , 38.        , 57.        ,
       40.5       ,  0.92      , 19.        , 36.5       ,  0.75

In [0]:
#5-1


train_cleaned = train.drop(["Name", "Ticket"], axis =1) 
train_cleaned.isnull()
train_cleaned.head()



PassengerId  Survived  Pclass  Sex  ...  Parch     Fare  Cabin  Embarked
0            1         0       3    1  ...      0   7.2500      0         2
1            2         1       1    0  ...      0  71.2833      1         0
2            3         1       3    0  ...      0   7.9250      0         2
3            4         1       1    0  ...      0  53.1000      1         2
4            5         0       3    1  ...      0   8.0500      0         2

[5 rows x 10 columns]

In [0]:
#attribuer le bon type aux données
train["Sex"] = pd.Categorical(train["Sex"], ordered = False)
train["Cabin"] = pd.Categorical(train["Cabin"], ordered = False)
train["Embarked"] = pd.Categorical(train["Embarked"], ordered = False)
train.head()


NameError: ignored

In [0]:
train_cleaned.to_csv('train_cleaned.csv', index=False)
files.download("train_cleaned.csv")

In [0]:
test_cleaned = test.drop([ "Name", "Ticket"], axis =1) 
test_cleaned.to_csv('test_cleaned.csv', index = False)
files.download("test_cleaned.csv")